resim/math/newton_solver.hh

```diff
@@ -0,0 +1,72 @@
+// Copyright 2024 ReSim, Inc.
+//
+// Use of this source code is governed by an MIT-style
+// license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+#pragma once
+
+#include <Eigen/Dense>
+
+#include "resim/utils/nullable_reference.hh"
+#include "resim/utils/status_value.hh"
+
+namespace resim::math {
+
+template <int DIM>
+using DifferentiableFunction = std::function<Eigen::Matrix<double, DIM, 1>(
+    const Eigen::Matrix<double, DIM, 1> &,
+    NullableReference<Eigen::Matrix<double, DIM, DIM>>)>;
+
+template <int DIM>
+StatusValue<Eigen::Matrix<double, DIM, 1>> newton_solve(
+    const DifferentiableFunction<DIM> &fun,
+    const Eigen::Matrix<double, DIM, 1> &guess,
+    const int max_iterations,
+    const double tolerance) {
+  using Vec = Eigen::Matrix<double, DIM, 1>;
+  using Mat = Eigen::Matrix<double, DIM, DIM>;
+  Vec result{guess};
+  Mat jacobian{Mat::Zero()};
+  for (int ii = 0; ii <= max_iterations; ++ii) {
+    const Vec error{fun(result, NullableReference{jacobian})};
+    if (error.norm() < tolerance) {
+      return result;
+    }
+    Eigen::JacobiSVD<Mat> svd{
+        jacobian,
+        Eigen::ComputeFullU | Eigen::ComputeFullV};
+    result = result - svd.solve(error);
+  }
+  return MAKE_STATUS("Did not converge in Newton iteration!");
+}
+
+template <typename Callable, typename Derived>
+StatusValue<Eigen::Matrix<double, Derived::RowsAtCompileTime, 1>> newton_solve(
```

Comment:
Could you comment the reason for the wrapper function here? Other than being able to chuck in lambdas without converting them to differentiable functions, it seems the main thing it achieves is the `static_assert` for catching lazy `auto` usage. But would that not be caught by the construction `DifferentiableFunction<DIM>(fun)` anyway, given this has an explicit output type? (Could be wrong on that, but that's why I want a little more clarity.) 

Happy if this is just a nice piece of syntactic sugar, just want the reasoning/usage to be clear to users.

Reply:
Yeah. This is a bit subtle, but it's to allow lambdas to be passed in directly here. I've added some documentation on this.